numpyro/distributions/util.py

```diff
@@ -13,21 +13,146 @@
+    """
+    Based on the transformed rejection sampling algorithm (BTRS) from the
+    following reference:
+
+    Hormann, "The Generation of Binonmial Random Variates"
+    (https://core.ac.uk/download/pdf/11007254.pdf)
+    """
+    # reshape to map over axis 0
+    tr_params = tree_map(lambda x: np.reshape(np.broadcast_to(x, shape), -1), _get_tr_params(n, p))
+    p = np.reshape(np.broadcast_to(p, shape), -1)
+    n = np.reshape(np.broadcast_to(n, shape), -1)
+    cont = np.reshape(np.broadcast_to(cont, shape), -1)
+    key = random.split(key, np.size(p))
+    # init values for (k, u, v) satisfies _btrs_cond_fn
+    ret = lax.map(lambda init_val: lax.while_loop(_btrs_cond_fn, _btrs_body_fn, init_val),
+                  (n + np.ones_like(n), key, p, n, tr_params, np.ones_like(p), np.ones_like(p), cont))
+    return np.reshape(ret[0], shape)
+
+
+def _binom_inv_body_fn(val):
+    i, key, p, n, geom_acc, cont = val
+    key, key_u = random.split(key)
+    u = random.uniform(key_u)
+    geom = np.clip(np.ceil(np.log(1 - u) / np.log(1 - p)), a_min=1., a_max=n + 1.)
+    geom_acc += geom
+    return i + 1, key, p, n, geom_acc, cont
+
+
+def _binom_inv_cond_fn(val):
+    i, _, _, n, geom_acc, cont = val
+    return cont & (geom_acc <= n)
+
+
+def _binomial_inversion(key, p, n, shape, cont):
+    # reshape to map over axis 0
+    p = np.reshape(np.broadcast_to(p, shape), -1)
+    n = np.reshape(np.broadcast_to(n, shape), -1)
+    cont = np.reshape(np.broadcast_to(cont, shape), -1)
+    key = random.split(key, np.size(p))
+    ret = lax.map(lambda init_val: lax.while_loop(_binom_inv_cond_fn, _binom_inv_body_fn, init_val),
+                  (np.zeros_like(n), key, p, n, np.zeros_like(p), cont))
+    return np.reshape(ret[0] - 1, shape)
+
+
+@partial(jit, static_argnums=(3,))
+def _binomial(key, p, n, shape):
     p, n = promote_shapes(p, n)
     shape = shape or lax.broadcast_shapes(np.shape(p), np.shape(n))
-    uniforms = random.uniform(key, shape + (n_max,))
-    n = np.expand_dims(n, axis=-1)
-    p = np.expand_dims(p, axis=-1)
-    mask = (np.arange(n_max) < n).astype(uniforms.dtype)
-    p, uniforms = promote_shapes(p, uniforms)
-    return np.sum(mask * lax.lt(uniforms, p), axis=-1, keepdims=False)
+    eps = 1e-6
+    q = 1 - p
+    idx_zeros = (p <= 0) | (n == 0)
+    idx_ones = (p >= 1)
+    idx_nans = np.isnan(p)
+    idx_le_mid = p <= 0.5
+    pq = np.where(idx_le_mid, p, q)
+    mu = n * pq
+    ret = np.zeros(shape, dtype=get_dtype(n))
+    # pre-fill for p = (0, 1, nan) as these values are problematic
+    ret = np.where(idx_zeros, 0, ret)
+    ret = np.where(idx_ones, n, ret)
+    ret = np.where(idx_nans, float('NaN'), ret)
+    p = np.clip(p, a_min=eps, a_max=1 - eps)
+    cont = np.logical_not(idx_zeros | idx_ones | idx_nans)
+    ret = np.where(cont & (mu <= 10), _binomial_inversion(key, p, n, shape, cont & (mu <= 10)), ret)
+    ret = np.where(cont & (mu > 10), _binomial_btrs(key, pq, n, shape, cont & (mu > 10)), ret)
```

Comment:
I think it is better to have `_binomial_one` implemented with a `lax.cond` defer to either `_binomial_inversion` or `_binomial_btrs`. Then using `lax.map` or `lax.vmap` here.

Reply:
You are right, I think that will be a better way to structure this and will likely simplify some of the code for early termination using `np.where`. I'll make the change.

Reply:
The code is much cleaner, thanks to your suggestion. There is a minor overhead associated with computing parameters like `mu`, `tr_params` etc. on the broadcasted tensor instead of the original unbroadcasted tensor as I was doing earlier. However, the alternative of computing all of these and passing it around will complicate the code, and I'm not sure if its worth the perf benefit.